// osmojs/main/codegen/osmosis/protorev/v1beta1/tx.rpc.msg.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MsgClientImpl = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _m0 = _interopRequireWildcard(require("protobufjs/minimal"));
var _tx = require("./tx");
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
var MsgClientImpl = /*#__PURE__*/function () {
  function MsgClientImpl(rpc) {
    (0, _classCallCheck2["default"])(this, MsgClientImpl);
    (0, _defineProperty2["default"])(this, "rpc", void 0);
    this.rpc = rpc;
    this.setHotRoutes = this.setHotRoutes.bind(this);
    this.setDeveloperAccount = this.setDeveloperAccount.bind(this);
    this.setMaxPoolPointsPerTx = this.setMaxPoolPointsPerTx.bind(this);
    this.setMaxPoolPointsPerBlock = this.setMaxPoolPointsPerBlock.bind(this);
    this.setPoolWeights = this.setPoolWeights.bind(this);
    this.setBaseDenoms = this.setBaseDenoms.bind(this);
  }
  (0, _createClass2["default"])(MsgClientImpl, [{
    key: "setHotRoutes",
    value: function setHotRoutes(request) {
      var data = _tx.MsgSetHotRoutes.encode(request).finish();
      var promise = this.rpc.request("osmosis.protorev.v1beta1.Msg", "SetHotRoutes", data);
      return promise.then(function (data) {
        return _tx.MsgSetHotRoutesResponse.decode(new _m0.Reader(data));
      });
    }
  }, {
    key: "setDeveloperAccount",
    value: function setDeveloperAccount(request) {
      var data = _tx.MsgSetDeveloperAccount.encode(request).finish();
      var promise = this.rpc.request("osmosis.protorev.v1beta1.Msg", "SetDeveloperAccount", data);
      return promise.then(function (data) {
        return _tx.MsgSetDeveloperAccountResponse.decode(new _m0.Reader(data));
      });
    }
  }, {
    key: "setMaxPoolPointsPerTx",
    value: function setMaxPoolPointsPerTx(request) {
      var data = _tx.MsgSetMaxPoolPointsPerTx.encode(request).finish();
      var promise = this.rpc.request("osmosis.protorev.v1beta1.Msg", "SetMaxPoolPointsPerTx", data);
      return promise.then(function (data) {
        return _tx.MsgSetMaxPoolPointsPerTxResponse.decode(new _m0.Reader(data));
      });
    }
  }, {
    key: "setMaxPoolPointsPerBlock",
    value: function setMaxPoolPointsPerBlock(request) {
      var data = _tx.MsgSetMaxPoolPointsPerBlock.encode(request).finish();
      var promise = this.rpc.request("osmosis.protorev.v1beta1.Msg", "SetMaxPoolPointsPerBlock", data);
      return promise.then(function (data) {
        return _tx.MsgSetMaxPoolPointsPerBlockResponse.decode(new _m0.Reader(data));
      });
    }
  }, {
    key: "setPoolWeights",
    value: function setPoolWeights(request) {
      var data = _tx.MsgSetPoolWeights.encode(request).finish();
      var promise = this.rpc.request("osmosis.protorev.v1beta1.Msg", "SetPoolWeights", data);
      return promise.then(function (data) {
        return _tx.MsgSetPoolWeightsResponse.decode(new _m0.Reader(data));
      });
    }
  }, {
    key: "setBaseDenoms",
    value: function setBaseDenoms(request) {
      var data = _tx.MsgSetBaseDenoms.encode(request).finish();
      var promise = this.rpc.request("osmosis.protorev.v1beta1.Msg", "SetBaseDenoms", data);
      return promise.then(function (data) {
        return _tx.MsgSetBaseDenomsResponse.decode(new _m0.Reader(data));
      });
    }
  }]);
  return MsgClientImpl;
}();
exports.MsgClientImpl = MsgClientImpl;
