// osmojs/main/codegen/osmosis/concentrated-liquidity/pool-model/query.rpc.Query.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createRpcQueryExtension = exports.QueryClientImpl = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _m0 = _interopRequireWildcard(require("protobufjs/minimal"));
var _stargate = require("@cosmjs/stargate");
var _query = require("./query");
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
var QueryClientImpl = /*#__PURE__*/function () {
  function QueryClientImpl(rpc) {
    (0, _classCallCheck2["default"])(this, QueryClientImpl);
    (0, _defineProperty2["default"])(this, "rpc", void 0);
    this.rpc = rpc;
    this.pools = this.pools.bind(this);
    this.params = this.params.bind(this);
    this.userPositions = this.userPositions.bind(this);
    this.totalLiquidityForRange = this.totalLiquidityForRange.bind(this);
    this.liquidityNetInDirection = this.liquidityNetInDirection.bind(this);
    this.claimableFees = this.claimableFees.bind(this);
    this.positionById = this.positionById.bind(this);
  }
  (0, _createClass2["default"])(QueryClientImpl, [{
    key: "pools",
    value: function pools() {
      var request = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {
        pagination: undefined
      };
      var data = _query.QueryPoolsRequest.encode(request).finish();
      var promise = this.rpc.request("osmosis.concentratedliquidity.v1beta1.Query", "Pools", data);
      return promise.then(function (data) {
        return _query.QueryPoolsResponse.decode(new _m0.Reader(data));
      });
    }
  }, {
    key: "params",
    value: function params() {
      var request = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      var data = _query.QueryParamsRequest.encode(request).finish();
      var promise = this.rpc.request("osmosis.concentratedliquidity.v1beta1.Query", "Params", data);
      return promise.then(function (data) {
        return _query.QueryParamsResponse.decode(new _m0.Reader(data));
      });
    }
  }, {
    key: "userPositions",
    value: function userPositions(request) {
      var data = _query.QueryUserPositionsRequest.encode(request).finish();
      var promise = this.rpc.request("osmosis.concentratedliquidity.v1beta1.Query", "UserPositions", data);
      return promise.then(function (data) {
        return _query.QueryUserPositionsResponse.decode(new _m0.Reader(data));
      });
    }
  }, {
    key: "totalLiquidityForRange",
    value: function totalLiquidityForRange(request) {
      var data = _query.QueryTotalLiquidityForRangeRequest.encode(request).finish();
      var promise = this.rpc.request("osmosis.concentratedliquidity.v1beta1.Query", "TotalLiquidityForRange", data);
      return promise.then(function (data) {
        return _query.QueryTotalLiquidityForRangeResponse.decode(new _m0.Reader(data));
      });
    }
  }, {
    key: "liquidityNetInDirection",
    value: function liquidityNetInDirection(request) {
      var data = _query.QueryLiquidityNetInDirectionRequest.encode(request).finish();
      var promise = this.rpc.request("osmosis.concentratedliquidity.v1beta1.Query", "LiquidityNetInDirection", data);
      return promise.then(function (data) {
        return _query.QueryLiquidityNetInDirectionResponse.decode(new _m0.Reader(data));
      });
    }
  }, {
    key: "claimableFees",
    value: function claimableFees(request) {
      var data = _query.QueryClaimableFeesRequest.encode(request).finish();
      var promise = this.rpc.request("osmosis.concentratedliquidity.v1beta1.Query", "ClaimableFees", data);
      return promise.then(function (data) {
        return _query.QueryClaimableFeesResponse.decode(new _m0.Reader(data));
      });
    }
  }, {
    key: "positionById",
    value: function positionById(request) {
      var data = _query.QueryPositionByIdRequest.encode(request).finish();
      var promise = this.rpc.request("osmosis.concentratedliquidity.v1beta1.Query", "PositionById", data);
      return promise.then(function (data) {
        return _query.QueryPositionByIdResponse.decode(new _m0.Reader(data));
      });
    }
  }]);
  return QueryClientImpl;
}();
exports.QueryClientImpl = QueryClientImpl;
var createRpcQueryExtension = function createRpcQueryExtension(base) {
  var rpc = (0, _stargate.createProtobufRpcClient)(base);
  var queryService = new QueryClientImpl(rpc);
  return {
    pools: function pools(request) {
      return queryService.pools(request);
    },
    params: function params(request) {
      return queryService.params(request);
    },
    userPositions: function userPositions(request) {
      return queryService.userPositions(request);
    },
    totalLiquidityForRange: function totalLiquidityForRange(request) {
      return queryService.totalLiquidityForRange(request);
    },
    liquidityNetInDirection: function liquidityNetInDirection(request) {
      return queryService.liquidityNetInDirection(request);
    },
    claimableFees: function claimableFees(request) {
      return queryService.claimableFees(request);
    },
    positionById: function positionById(request) {
      return queryService.positionById(request);
    }
  };
};
exports.createRpcQueryExtension = createRpcQueryExtension;
