// osmojs/module/codegen/osmosis/tokenfactory/v1beta1/authorityMetadata.js
import * as _m0 from "protobufjs/minimal";
/**
 * DenomAuthorityMetadata specifies metadata for addresses that have specific
 * capabilities over a token factory denom. Right now there is only one Admin
 * permission, but is planned to be extended to the future.
 */

function createBaseDenomAuthorityMetadata() {
  return {
    admin: ""
  };
}
export const DenomAuthorityMetadata = {
  typeUrl: "/osmosis.tokenfactory.v1beta1.DenomAuthorityMetadata",
  encode(message, writer = _m0.Writer.create()) {
    if (message.admin !== "") {
      writer.uint32(10).string(message.admin);
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseDenomAuthorityMetadata();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          message.admin = reader.string();
          break;
        default:
          reader.skipType(tag & 7);
          break;
      }
    }
    return message;
  },
  fromPartial(object) {
    var _object$admin;
    const message = createBaseDenomAuthorityMetadata();
    message.admin = (_object$admin = object.admin) !== null && _object$admin !== void 0 ? _object$admin : "";
    return message;
  },
  fromAmino(object) {
    return {
      admin: object.admin
    };
  },
  toAmino(message) {
    const obj = {};
    obj.admin = message.admin;
    return obj;
  },
  fromAminoMsg(object) {
    return DenomAuthorityMetadata.fromAmino(object.value);
  },
  toAminoMsg(message) {
    return {
      type: "osmosis/tokenfactory/denom-authority-metadata",
      value: DenomAuthorityMetadata.toAmino(message)
    };
  },
  fromProtoMsg(message) {
    return DenomAuthorityMetadata.decode(message.value);
  },
  toProto(message) {
    return DenomAuthorityMetadata.encode(message).finish();
  },
  toProtoMsg(message) {
    return {
      typeUrl: "/osmosis.tokenfactory.v1beta1.DenomAuthorityMetadata",
      value: DenomAuthorityMetadata.encode(message).finish()
    };
  }
};
