// @cosmology/core/main/clients/cosmos.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CosmosApiClient = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _rest = require("./rest");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var CosmosApiClient = /*#__PURE__*/function (_RestClient) {
  (0, _inherits2["default"])(CosmosApiClient, _RestClient);

  var _super = _createSuper(CosmosApiClient);

  function CosmosApiClient(_ref) {
    var _this;

    var url = _ref.url;
    (0, _classCallCheck2["default"])(this, CosmosApiClient);
    _this = _super.call(this, {
      url: url
    });
    _this._clientType = 'Cosmos API';
    return _this;
  }

  (0, _createClass2["default"])(CosmosApiClient, [{
    key: "getTransaction",
    value: function () {
      var _getTransaction = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(txHash) {
        var endpoint;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                endpoint = "txs/".concat(txHash);
                _context.next = 3;
                return this.request(endpoint);

              case 3:
                return _context.abrupt("return", _context.sent);

              case 4:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function getTransaction(_x) {
        return _getTransaction.apply(this, arguments);
      }

      return getTransaction;
    }()
    /**
     * @returns {Promise<{
     *   tx: {
     *      body: object;
     *      authInfo: object;
     *      signatures: object[];
     *   };
     *   tx_response: {
     *      height: string;
     *      txhash: string;
     *   }
     * }>}
     */

  }, {
    key: "getCosmosTransaction",
    value: function () {
      var _getCosmosTransaction = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2(txHash) {
        var endpoint;
        return _regenerator["default"].wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                endpoint = "cosmos/tx/v1beta1/txs/".concat(txHash);
                _context2.next = 3;
                return this.request(endpoint);

              case 3:
                return _context2.abrupt("return", _context2.sent);

              case 4:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function getCosmosTransaction(_x2) {
        return _getCosmosTransaction.apply(this, arguments);
      }

      return getCosmosTransaction;
    }()
  }, {
    key: "getProposals",
    value: function () {
      var _getProposals = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3() {
        var proposalStatus,
            endpoint,
            _args3 = arguments;
        return _regenerator["default"].wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                proposalStatus = _args3.length > 0 && _args3[0] !== undefined ? _args3[0] : 2;
                endpoint = "cosmos/gov/v1beta1/proposals";

                if (!proposalStatus) {
                  _context3.next = 6;
                  break;
                }

                _context3.next = 5;
                return this.request(endpoint, {
                  params: {
                    proposal_status: proposalStatus
                  }
                });

              case 5:
                return _context3.abrupt("return", _context3.sent);

              case 6:
                _context3.next = 8;
                return this.request(endpoint);

              case 8:
                return _context3.abrupt("return", _context3.sent);

              case 9:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function getProposals() {
        return _getProposals.apply(this, arguments);
      }

      return getProposals;
    }()
    /**
     * @returns {Promise<{
     *   block_id: {
     *      hash: string;
     *      part_set_header: object;
     *   };
     *   block: {
     *      header: {
     *        version: object;
     *        chain_id: string;
     *        height: string;
     *        time: string;
     *      }
     *   }
     * }>}
     */

  }, {
    key: "getLatestBlock",
    value: function () {
      var _getLatestBlock = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee4() {
        var endpoint;
        return _regenerator["default"].wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                endpoint = "/cosmos/base/tendermint/v1beta1/blocks/latest";
                _context4.next = 3;
                return this.request(endpoint);

              case 3:
                return _context4.abrupt("return", _context4.sent);

              case 4:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      function getLatestBlock() {
        return _getLatestBlock.apply(this, arguments);
      }

      return getLatestBlock;
    }() // below is the subset only supported by keplr.app endpoint...

  }, {
    key: "getBalances",
    value: function () {
      var _getBalances = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee5(address) {
        var endpoint;
        return _regenerator["default"].wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                endpoint = "bank/balances/".concat(address);
                _context5.next = 3;
                return this.request(endpoint);

              case 3:
                return _context5.abrupt("return", _context5.sent);

              case 4:
              case "end":
                return _context5.stop();
            }
          }
        }, _callee5, this);
      }));

      function getBalances(_x3) {
        return _getBalances.apply(this, arguments);
      }

      return getBalances;
    }()
  }, {
    key: "authInfo",
    value: function () {
      var _authInfo = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee6(address) {
        var endpoint;
        return _regenerator["default"].wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                endpoint = "auth/accounts/".concat(address);
                _context6.next = 3;
                return this.request(endpoint);

              case 3:
                return _context6.abrupt("return", _context6.sent);

              case 4:
              case "end":
                return _context6.stop();
            }
          }
        }, _callee6, this);
      }));

      function authInfo(_x4) {
        return _authInfo.apply(this, arguments);
      }

      return authInfo;
    }()
  }, {
    key: "getUnbondingDelegations",
    value: function () {
      var _getUnbondingDelegations = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee7(address) {
        var endpoint;
        return _regenerator["default"].wrap(function _callee7$(_context7) {
          while (1) {
            switch (_context7.prev = _context7.next) {
              case 0:
                endpoint = "staking/delegators/".concat(address, "/unbonding_delegations");
                _context7.next = 3;
                return this.request(endpoint);

              case 3:
                return _context7.abrupt("return", _context7.sent);

              case 4:
              case "end":
                return _context7.stop();
            }
          }
        }, _callee7, this);
      }));

      function getUnbondingDelegations(_x5) {
        return _getUnbondingDelegations.apply(this, arguments);
      }

      return getUnbondingDelegations;
    }()
  }, {
    key: "getDelegations",
    value: function () {
      var _getDelegations = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee8(address) {
        var endpoint;
        return _regenerator["default"].wrap(function _callee8$(_context8) {
          while (1) {
            switch (_context8.prev = _context8.next) {
              case 0:
                endpoint = "staking/delegators/".concat(address, "/delegations");
                _context8.next = 3;
                return this.request(endpoint);

              case 3:
                return _context8.abrupt("return", _context8.sent);

              case 4:
              case "end":
                return _context8.stop();
            }
          }
        }, _callee8, this);
      }));

      function getDelegations(_x6) {
        return _getDelegations.apply(this, arguments);
      }

      return getDelegations;
    }()
  }, {
    key: "getDelegatorValidatorInfo",
    value: function () {
      var _getDelegatorValidatorInfo = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee9(delegatorAddr, validatorAddr) {
        var endpoint;
        return _regenerator["default"].wrap(function _callee9$(_context9) {
          while (1) {
            switch (_context9.prev = _context9.next) {
              case 0:
                endpoint = "cosmos/staking/v1beta1/delegators/".concat(delegatorAddr, "/validators/").concat(validatorAddr);
                _context9.next = 3;
                return this.request(endpoint);

              case 3:
                return _context9.abrupt("return", _context9.sent);

              case 4:
              case "end":
                return _context9.stop();
            }
          }
        }, _callee9, this);
      }));

      function getDelegatorValidatorInfo(_x7, _x8) {
        return _getDelegatorValidatorInfo.apply(this, arguments);
      }

      return getDelegatorValidatorInfo;
    }()
  }, {
    key: "getValidatorInfo",
    value: function () {
      var _getValidatorInfo = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee10(validatorAddr) {
        var endpoint;
        return _regenerator["default"].wrap(function _callee10$(_context10) {
          while (1) {
            switch (_context10.prev = _context10.next) {
              case 0:
                endpoint = "cosmos/staking/v1beta1/validators/".concat(validatorAddr);
                _context10.next = 3;
                return this.request(endpoint);

              case 3:
                return _context10.abrupt("return", _context10.sent);

              case 4:
              case "end":
                return _context10.stop();
            }
          }
        }, _callee10, this);
      }));

      function getValidatorInfo(_x9) {
        return _getValidatorInfo.apply(this, arguments);
      }

      return getValidatorInfo;
    }()
  }, {
    key: "getValidators",
    value: function () {
      var _getValidators = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee11(status) {
        var endpoint;
        return _regenerator["default"].wrap(function _callee11$(_context11) {
          while (1) {
            switch (_context11.prev = _context11.next) {
              case 0:
                endpoint = "cosmos/staking/v1beta1/validators";

                if (!status) {
                  _context11.next = 5;
                  break;
                }

                _context11.next = 4;
                return this.request(endpoint, {
                  params: {
                    status: status
                  }
                });

              case 4:
                return _context11.abrupt("return", _context11.sent);

              case 5:
                _context11.next = 7;
                return this.request(endpoint);

              case 7:
                return _context11.abrupt("return", _context11.sent);

              case 8:
              case "end":
                return _context11.stop();
            }
          }
        }, _callee11, this);
      }));

      function getValidators(_x10) {
        return _getValidators.apply(this, arguments);
      }

      return getValidators;
    }()
  }, {
    key: "getRewards",
    value: function () {
      var _getRewards = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee12(address) {
        var endpoint;
        return _regenerator["default"].wrap(function _callee12$(_context12) {
          while (1) {
            switch (_context12.prev = _context12.next) {
              case 0:
                endpoint = "cosmos/distribution/v1beta1/delegators/".concat(address, "/rewards");
                _context12.next = 3;
                return this.request(endpoint);

              case 3:
                return _context12.abrupt("return", _context12.sent);

              case 4:
              case "end":
                return _context12.stop();
            }
          }
        }, _callee12, this);
      }));

      function getRewards(_x11) {
        return _getRewards.apply(this, arguments);
      }

      return getRewards;
    }()
  }]);
  return CosmosApiClient;
}(_rest.RestClient);

exports.CosmosApiClient = CosmosApiClient;
