// @cosmology/core/main/clients/osmosis.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.prettyPool = exports.OsmosisApiClient = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assets = require("../assets");

var _cosmos = require("./cosmos");

var _classAutobind = _interopRequireDefault(require("class-autobind"));

var _unit = require("@keplr-wallet/unit");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var assetHashMap = _assets.assets.reduce(function (m, asset) {
  m[asset.base] = asset;
  return m;
}, {}); // Cosmos LCD API
// https://www.notion.so/Stake-Systems-LCD-RPC-gRPC-Instances-04a99a9a9aa14247a42944931eec7024


var OsmosisApiClient = /*#__PURE__*/function (_CosmosApiClient) {
  (0, _inherits2["default"])(OsmosisApiClient, _CosmosApiClient);

  var _super = _createSuper(OsmosisApiClient);

  function OsmosisApiClient() {
    var _this;

    var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
        _ref$url = _ref.url,
        url = _ref$url === void 0 ? 'https://osmosis.stakesystems.io/' : _ref$url;

    (0, _classCallCheck2["default"])(this, OsmosisApiClient);
    _this = _super.call(this, {
      url: url
    });
    _this._clientType = 'Osmosis API';
    (0, _classAutobind["default"])((0, _assertThisInitialized2["default"])(_this)); // React ES6 doesn't bind this -> meaning we get 'unable to read property 'request' of undefined

    return _this;
  }

  (0, _createClass2["default"])(OsmosisApiClient, [{
    key: "getPools",
    value: function () {
      var _getPools = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee() {
        var endpoint;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                endpoint = "osmosis/gamm/v1beta1/pools?pagination.limit=2000";
                _context.next = 3;
                return this.request(endpoint, {
                  'Cache-Control': 'max-age=60'
                });

              case 3:
                return _context.abrupt("return", _context.sent);

              case 4:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function getPools() {
        return _getPools.apply(this, arguments);
      }

      return getPools;
    }()
  }, {
    key: "getPool",
    value: function () {
      var _getPool = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2(poolId) {
        var endpoint;
        return _regenerator["default"].wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                endpoint = "osmosis/gamm/v1beta1/pools/".concat(poolId);
                _context2.next = 3;
                return this.request(endpoint);

              case 3:
                return _context2.abrupt("return", _context2.sent);

              case 4:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function getPool(_x) {
        return _getPool.apply(this, arguments);
      }

      return getPool;
    }()
  }, {
    key: "getAccountLockedLongerDuration",
    value: function () {
      var _getAccountLockedLongerDuration = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3(address) {
        var endpoint;
        return _regenerator["default"].wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                endpoint = "osmosis/lockup/v1beta1/account_locked_longer_duration/".concat(address);
                _context3.next = 3;
                return this.request(endpoint);

              case 3:
                return _context3.abrupt("return", _context3.sent);

              case 4:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function getAccountLockedLongerDuration(_x2) {
        return _getAccountLockedLongerDuration.apply(this, arguments);
      }

      return getAccountLockedLongerDuration;
    }()
  }, {
    key: "getAccountLockedCoins",
    value: function () {
      var _getAccountLockedCoins = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee4(address) {
        var endpoint;
        return _regenerator["default"].wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                endpoint = "osmosis/lockup/v1beta1/account_locked_coins/".concat(address);
                _context4.next = 3;
                return this.request(endpoint);

              case 3:
                return _context4.abrupt("return", _context4.sent);

              case 4:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      function getAccountLockedCoins(_x3) {
        return _getAccountLockedCoins.apply(this, arguments);
      }

      return getAccountLockedCoins;
    }()
  }, {
    key: "getEpochProvision",
    value: function () {
      var _getEpochProvision = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee5() {
        var endpoint;
        return _regenerator["default"].wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                endpoint = "osmosis/mint/v1beta1/epoch_provisions";
                _context5.next = 3;
                return this.request(endpoint);

              case 3:
                return _context5.abrupt("return", _context5.sent);

              case 4:
              case "end":
                return _context5.stop();
            }
          }
        }, _callee5, this);
      }));

      function getEpochProvision() {
        return _getEpochProvision.apply(this, arguments);
      }

      return getEpochProvision;
    }()
  }, {
    key: "getEpochs",
    value: function () {
      var _getEpochs = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee6() {
        var endpoint;
        return _regenerator["default"].wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                endpoint = "osmosis/epochs/v1beta1/epochs";
                _context6.next = 3;
                return this.request(endpoint);

              case 3:
                return _context6.abrupt("return", _context6.sent);

              case 4:
              case "end":
                return _context6.stop();
            }
          }
        }, _callee6, this);
      }));

      function getEpochs() {
        return _getEpochs.apply(this, arguments);
      }

      return getEpochs;
    }()
  }, {
    key: "getDistrInfo",
    value: function () {
      var _getDistrInfo = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee7() {
        var endpoint;
        return _regenerator["default"].wrap(function _callee7$(_context7) {
          while (1) {
            switch (_context7.prev = _context7.next) {
              case 0:
                endpoint = "osmosis/pool-incentives/v1beta1/distr_info";
                _context7.next = 3;
                return this.request(endpoint);

              case 3:
                return _context7.abrupt("return", _context7.sent);

              case 4:
              case "end":
                return _context7.stop();
            }
          }
        }, _callee7, this);
      }));

      function getDistrInfo() {
        return _getDistrInfo.apply(this, arguments);
      }

      return getDistrInfo;
    }()
  }, {
    key: "getParams",
    value: function () {
      var _getParams = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee8() {
        var endpoint;
        return _regenerator["default"].wrap(function _callee8$(_context8) {
          while (1) {
            switch (_context8.prev = _context8.next) {
              case 0:
                endpoint = "osmosis/mint/v1beta1/params";
                _context8.next = 3;
                return this.request(endpoint);

              case 3:
                return _context8.abrupt("return", _context8.sent);

              case 4:
              case "end":
                return _context8.stop();
            }
          }
        }, _callee8, this);
      }));

      function getParams() {
        return _getParams.apply(this, arguments);
      }

      return getParams;
    }()
  }, {
    key: "getLockableDurations",
    value: function () {
      var _getLockableDurations = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee9() {
        var endpoint;
        return _regenerator["default"].wrap(function _callee9$(_context9) {
          while (1) {
            switch (_context9.prev = _context9.next) {
              case 0:
                endpoint = "osmosis/pool-incentives/v1beta1/lockable_durations";
                _context9.next = 3;
                return this.request(endpoint);

              case 3:
                return _context9.abrupt("return", _context9.sent);

              case 4:
              case "end":
                return _context9.stop();
            }
          }
        }, _callee9, this);
      }));

      function getLockableDurations() {
        return _getLockableDurations.apply(this, arguments);
      }

      return getLockableDurations;
    }() // https://osmosis.stakesystems.io/osmosis/incentives/v1beta1/gauges
    // returns gauges both upcoming and active

  }, {
    key: "getGauges",
    value: function () {
      var _getGauges = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee10() {
        var endpoint;
        return _regenerator["default"].wrap(function _callee10$(_context10) {
          while (1) {
            switch (_context10.prev = _context10.next) {
              case 0:
                endpoint = "osmosis/incentives/v1beta1/gauges";
                _context10.next = 3;
                return this.request(endpoint);

              case 3:
                return _context10.abrupt("return", _context10.sent);

              case 4:
              case "end":
                return _context10.stop();
            }
          }
        }, _callee10, this);
      }));

      function getGauges() {
        return _getGauges.apply(this, arguments);
      }

      return getGauges;
    }()
  }, {
    key: "getActiveGauges",
    value: function () {
      var _getActiveGauges = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee11() {
        var endpoint;
        return _regenerator["default"].wrap(function _callee11$(_context11) {
          while (1) {
            switch (_context11.prev = _context11.next) {
              case 0:
                endpoint = "osmosis/incentives/v1beta1/active_gauges";
                _context11.next = 3;
                return this.request(endpoint);

              case 3:
                return _context11.abrupt("return", _context11.sent);

              case 4:
              case "end":
                return _context11.stop();
            }
          }
        }, _callee11, this);
      }));

      function getActiveGauges() {
        return _getActiveGauges.apply(this, arguments);
      }

      return getActiveGauges;
    }() // https://osmosis.stakesystems.io/osmosis/incentives/v1beta1/gauge_by_id/5

  }, {
    key: "getGauge",
    value: function () {
      var _getGauge = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee12(gaugeId) {
        var endpoint;
        return _regenerator["default"].wrap(function _callee12$(_context12) {
          while (1) {
            switch (_context12.prev = _context12.next) {
              case 0:
                endpoint = "osmosis/incentives/v1beta1/gauge_by_id/".concat(gaugeId);
                _context12.next = 3;
                return this.request(endpoint);

              case 3:
                return _context12.abrupt("return", _context12.sent);

              case 4:
              case "end":
                return _context12.stop();
            }
          }
        }, _callee12, this);
      }));

      function getGauge(_x4) {
        return _getGauge.apply(this, arguments);
      }

      return getGauge;
    }() // https://osmosis.stakesystems.io/osmosis/pool-incentives/v1beta1/incentivized_pools

  }, {
    key: "getIncentivizedPools",
    value: function () {
      var _getIncentivizedPools = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee13() {
        var endpoint;
        return _regenerator["default"].wrap(function _callee13$(_context13) {
          while (1) {
            switch (_context13.prev = _context13.next) {
              case 0:
                endpoint = "osmosis/pool-incentives/v1beta1/incentivized_pools";
                _context13.next = 3;
                return this.request(endpoint);

              case 3:
                return _context13.abrupt("return", _context13.sent);

              case 4:
              case "end":
                return _context13.stop();
            }
          }
        }, _callee13, this);
      }));

      function getIncentivizedPools() {
        return _getIncentivizedPools.apply(this, arguments);
      }

      return getIncentivizedPools;
    }()
  }, {
    key: "getPoolsPretty",
    value: function () {
      var _getPoolsPretty = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee14() {
        var _ref2,
            _ref2$includeDetails,
            includeDetails,
            _yield$this$getPools,
            pools,
            _args14 = arguments;

        return _regenerator["default"].wrap(function _callee14$(_context14) {
          while (1) {
            switch (_context14.prev = _context14.next) {
              case 0:
                _ref2 = _args14.length > 0 && _args14[0] !== undefined ? _args14[0] : {}, _ref2$includeDetails = _ref2.includeDetails, includeDetails = _ref2$includeDetails === void 0 ? false : _ref2$includeDetails;
                _context14.next = 3;
                return this.getPools();

              case 3:
                _yield$this$getPools = _context14.sent;
                pools = _yield$this$getPools.pools;
                return _context14.abrupt("return", pools.map(function (pool) {
                  return prettyPool(pool, {
                    includeDetails: includeDetails
                  });
                }));

              case 6:
              case "end":
                return _context14.stop();
            }
          }
        }, _callee14, this);
      }));

      function getPoolsPretty() {
        return _getPoolsPretty.apply(this, arguments);
      }

      return getPoolsPretty;
    }()
  }, {
    key: "getPoolPretty",
    value: function () {
      var _getPoolPretty = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee15(poodId) {
        var _ref3,
            _ref3$includeDetails,
            includeDetails,
            _yield$this$getPool,
            pool,
            _args15 = arguments;

        return _regenerator["default"].wrap(function _callee15$(_context15) {
          while (1) {
            switch (_context15.prev = _context15.next) {
              case 0:
                _ref3 = _args15.length > 1 && _args15[1] !== undefined ? _args15[1] : {}, _ref3$includeDetails = _ref3.includeDetails, includeDetails = _ref3$includeDetails === void 0 ? false : _ref3$includeDetails;
                _context15.next = 3;
                return this.getPool(poodId);

              case 3:
                _yield$this$getPool = _context15.sent;
                pool = _yield$this$getPool.pool;
                return _context15.abrupt("return", prettyPool(pool, {
                  includeDetails: includeDetails
                }));

              case 6:
              case "end":
                return _context15.stop();
            }
          }
        }, _callee15, this);
      }));

      function getPoolPretty(_x5) {
        return _getPoolPretty.apply(this, arguments);
      }

      return getPoolPretty;
    }()
  }]);
  return OsmosisApiClient;
}(_cosmos.CosmosApiClient);

exports.OsmosisApiClient = OsmosisApiClient;

var prettyPool = function prettyPool(pool) {
  var _ref4 = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {},
      _ref4$includeDetails = _ref4.includeDetails,
      includeDetails = _ref4$includeDetails === void 0 ? false : _ref4$includeDetails;

  var totalWeight = new _unit.Dec(pool.totalWeight);
  var tokens = pool.poolAssets.map(function (_ref5) {
    var _asset$symbol;

    var token = _ref5.token,
        weight = _ref5.weight;
    var asset = assetHashMap === null || assetHashMap === void 0 ? void 0 : assetHashMap[token.denom];
    var symbol = (_asset$symbol = asset === null || asset === void 0 ? void 0 : asset.symbol) !== null && _asset$symbol !== void 0 ? _asset$symbol : token.denom;
    var w = new _unit.Dec(weight);
    var ratio = w.quo(totalWeight).toString();
    var obj = {
      symbol: symbol,
      denom: token.denom,
      amount: token.amount,
      ratio: ratio,
      info: undefined
    };

    if (includeDetails) {
      obj.info = asset;
    }

    return obj;
  });
  var value = {
    nickname: tokens.map(function (t) {
      return t.symbol;
    }).join('/'),
    images: undefined
  };

  if (includeDetails) {
    value.images = tokens.map(function (t) {
      var _t$info;

      var imgs = t === null || t === void 0 ? void 0 : (_t$info = t.info) === null || _t$info === void 0 ? void 0 : _t$info.logo_URIs;

      if (imgs) {
        return {
          token: t.symbol,
          images: imgs
        };
      }
    }).filter(Boolean);
  }

  return _objectSpread(_objectSpread(_objectSpread({}, value), pool), {}, {
    poolAssetsPretty: tokens
  });
};

exports.prettyPool = prettyPool;
