// @cosmology/core/main/clients/validator.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.OsmosisValidatorClient = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _rest = require("./rest");

var _classAutobind = _interopRequireDefault(require("class-autobind"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

//  https://api-osmosis.imperator.co/swagger/#/
var OsmosisValidatorClient = /*#__PURE__*/function (_RestClient) {
  (0, _inherits2["default"])(OsmosisValidatorClient, _RestClient);

  var _super = _createSuper(OsmosisValidatorClient);

  function OsmosisValidatorClient() {
    var _this;

    var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
        _ref$url = _ref.url,
        url = _ref$url === void 0 ? 'https://api-osmosis.imperator.co/' : _ref$url;

    (0, _classCallCheck2["default"])(this, OsmosisValidatorClient);
    _this = _super.call(this, {
      url: url
    });
    _this._clientType = 'Osmosis Validator';
    (0, _classAutobind["default"])((0, _assertThisInitialized2["default"])(_this)); // React ES6 doesn't bind this -> meaning we get 'unable to read property 'request' of undefined

    return _this;
  }

  (0, _createClass2["default"])(OsmosisValidatorClient, [{
    key: "getPools",
    value: function () {
      var _getPools = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee() {
        var endpoint;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                endpoint = "search/v1/pools";
                _context.next = 3;
                return this.request(endpoint);

              case 3:
                return _context.abrupt("return", _context.sent);

              case 4:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function getPools() {
        return _getPools.apply(this, arguments);
      }

      return getPools;
    }()
  }, {
    key: "getTokens",
    value: function () {
      var _getTokens = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2() {
        var endpoint;
        return _regenerator["default"].wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                endpoint = "tokens/v2/all";
                _context2.next = 3;
                return this.request(endpoint);

              case 3:
                return _context2.abrupt("return", _context2.sent);

              case 4:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function getTokens() {
        return _getTokens.apply(this, arguments);
      }

      return getTokens;
    }()
  }, {
    key: "getToken",
    value: function () {
      var _getToken = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3(symbol) {
        var endpoint;
        return _regenerator["default"].wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                endpoint = "tokens/v2/".concat(symbol);
                _context3.next = 3;
                return this.request(endpoint);

              case 3:
                return _context3.abrupt("return", _context3.sent);

              case 4:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function getToken(_x) {
        return _getToken.apply(this, arguments);
      }

      return getToken;
    }()
  }, {
    key: "getTokenPrice",
    value: function () {
      var _getTokenPrice = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee4(symbol) {
        var endpoint;
        return _regenerator["default"].wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                endpoint = "tokens/v2/price/".concat(symbol);
                _context4.next = 3;
                return this.request(endpoint);

              case 3:
                return _context4.abrupt("return", _context4.sent);

              case 4:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      function getTokenPrice(_x2) {
        return _getTokenPrice.apply(this, arguments);
      }

      return getTokenPrice;
    }()
  }, {
    key: "getPoolAprs",
    value: function () {
      var _getPoolAprs = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee5() {
        var endpoint;
        return _regenerator["default"].wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                endpoint = "apr/v2/all";
                _context5.next = 3;
                return this.request(endpoint);

              case 3:
                return _context5.abrupt("return", _context5.sent);

              case 4:
              case "end":
                return _context5.stop();
            }
          }
        }, _callee5, this);
      }));

      function getPoolAprs() {
        return _getPoolAprs.apply(this, arguments);
      }

      return getPoolAprs;
    }()
  }, {
    key: "getPoolApr",
    value: function () {
      var _getPoolApr = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee6(symbol) {
        var endpoint;
        return _regenerator["default"].wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                endpoint = "apr/v2/".concat(symbol);
                _context6.next = 3;
                return this.request(endpoint);

              case 3:
                return _context6.abrupt("return", _context6.sent);

              case 4:
              case "end":
                return _context6.stop();
            }
          }
        }, _callee6, this);
      }));

      function getPoolApr(_x3) {
        return _getPoolApr.apply(this, arguments);
      }

      return getPoolApr;
    }()
  }, {
    key: "getPairsSummary",
    value: function () {
      var _getPairsSummary = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee7() {
        var endpoint;
        return _regenerator["default"].wrap(function _callee7$(_context7) {
          while (1) {
            switch (_context7.prev = _context7.next) {
              case 0:
                endpoint = "pairs/v1/summary";
                _context7.next = 3;
                return this.request(endpoint);

              case 3:
                return _context7.abrupt("return", _context7.sent);

              case 4:
              case "end":
                return _context7.stop();
            }
          }
        }, _callee7, this);
      }));

      function getPairsSummary() {
        return _getPairsSummary.apply(this, arguments);
      }

      return getPairsSummary;
    }()
  }]);
  return OsmosisValidatorClient;
}(_rest.RestClient);

exports.OsmosisValidatorClient = OsmosisValidatorClient;
