// osmojs/module/codegen/osmosis/superfluid/params.js
import * as _m0 from "protobufjs/minimal";
/** Params holds parameters for the superfluid module */

function createBaseParams() {
  return {
    minimumRiskFactor: ""
  };
}
export const Params = {
  typeUrl: "/osmosis.superfluid.Params",
  encode(message, writer = _m0.Writer.create()) {
    if (message.minimumRiskFactor !== "") {
      writer.uint32(10).string(message.minimumRiskFactor);
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseParams();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          message.minimumRiskFactor = reader.string();
          break;
        default:
          reader.skipType(tag & 7);
          break;
      }
    }
    return message;
  },
  fromPartial(object) {
    var _object$minimumRiskFa;
    const message = createBaseParams();
    message.minimumRiskFactor = (_object$minimumRiskFa = object.minimumRiskFactor) !== null && _object$minimumRiskFa !== void 0 ? _object$minimumRiskFa : "";
    return message;
  },
  fromAmino(object) {
    return {
      minimumRiskFactor: object.minimum_risk_factor
    };
  },
  toAmino(message) {
    const obj = {};
    obj.minimum_risk_factor = message.minimumRiskFactor;
    return obj;
  },
  fromAminoMsg(object) {
    return Params.fromAmino(object.value);
  },
  toAminoMsg(message) {
    return {
      type: "osmosis/params",
      value: Params.toAmino(message)
    };
  },
  fromProtoMsg(message) {
    return Params.decode(message.value);
  },
  toProto(message) {
    return Params.encode(message).finish();
  },
  toProtoMsg(message) {
    return {
      typeUrl: "/osmosis.superfluid.Params",
      value: Params.encode(message).finish()
    };
  }
};
