// @cosmology/core/main/messages/utils.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.estimateOsmoFee = void 0;
Object.defineProperty(exports, "getSigningOsmosisClient", {
  enumerable: true,
  get: function get() {
    return _osmojs.getSigningOsmosisClient;
  }
});
exports.signAndBroadcastBatch = exports.signAndBroadcast = exports.noDecimals = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _stargate = require("@cosmjs/stargate");

var _tx = require("cosmjs-types/cosmos/tx/v1beta1/tx");

var _unit = require("@keplr-wallet/unit");

var _osmojs = require("osmojs");

var signAndBroadcast = /*#__PURE__*/function () {
  var _ref2 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(_ref) {
    var client, chainId, address, msg, fee, _ref$memo, memo, _yield$client$getSequ, accountNumber, sequence, txRaw, txBytes;

    return _regenerator["default"].wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            client = _ref.client, chainId = _ref.chainId, address = _ref.address, msg = _ref.msg, fee = _ref.fee, _ref$memo = _ref.memo, memo = _ref$memo === void 0 ? '' : _ref$memo;
            _context.next = 3;
            return client.getSequence(address);

          case 3:
            _yield$client$getSequ = _context.sent;
            accountNumber = _yield$client$getSequ.accountNumber;
            sequence = _yield$client$getSequ.sequence;
            _context.next = 8;
            return client.sign(address, [msg], fee, memo, {
              accountNumber: accountNumber,
              sequence: sequence,
              chainId: chainId
            });

          case 8:
            txRaw = _context.sent;
            txBytes = _tx.TxRaw.encode(txRaw).finish();
            _context.next = 12;
            return client.broadcastTx(txBytes);

          case 12:
            return _context.abrupt("return", _context.sent);

          case 13:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));

  return function signAndBroadcast(_x) {
    return _ref2.apply(this, arguments);
  };
}();

exports.signAndBroadcast = signAndBroadcast;

var signAndBroadcastBatch = /*#__PURE__*/function () {
  var _ref4 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2(_ref3) {
    var client, chainId, address, msgs, fee, _ref3$memo, memo, _yield$client$getSequ2, accountNumber, sequence, txRaw, txBytes;

    return _regenerator["default"].wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            client = _ref3.client, chainId = _ref3.chainId, address = _ref3.address, msgs = _ref3.msgs, fee = _ref3.fee, _ref3$memo = _ref3.memo, memo = _ref3$memo === void 0 ? '' : _ref3$memo;
            _context2.next = 3;
            return client.getSequence(address);

          case 3:
            _yield$client$getSequ2 = _context2.sent;
            accountNumber = _yield$client$getSequ2.accountNumber;
            sequence = _yield$client$getSequ2.sequence;
            _context2.next = 8;
            return client.sign(address, msgs, fee, memo, {
              accountNumber: accountNumber,
              sequence: sequence,
              chainId: chainId
            });

          case 8:
            txRaw = _context2.sent;
            txBytes = _tx.TxRaw.encode(txRaw).finish();
            _context2.next = 12;
            return client.broadcastTx(txBytes);

          case 12:
            return _context2.abrupt("return", _context2.sent);

          case 13:
          case "end":
            return _context2.stop();
        }
      }
    }, _callee2);
  }));

  return function signAndBroadcastBatch(_x2) {
    return _ref4.apply(this, arguments);
  };
}();

exports.signAndBroadcastBatch = signAndBroadcastBatch;

var estimateOsmoFee = /*#__PURE__*/function () {
  var _ref5 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3(client, address, msgs, memo) {
    var gasPrice, gasEstimation, fee;
    return _regenerator["default"].wrap(function _callee3$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            gasPrice = _stargate.GasPrice.fromString('0.025uosmo');
            _context3.next = 3;
            return client.simulate(address, msgs, memo);

          case 3:
            gasEstimation = _context3.sent;
            fee = (0, _stargate.calculateFee)(Math.round(gasEstimation * 1.3), gasPrice);
            return _context3.abrupt("return", fee);

          case 6:
          case "end":
            return _context3.stop();
        }
      }
    }, _callee3);
  }));

  return function estimateOsmoFee(_x3, _x4, _x5, _x6) {
    return _ref5.apply(this, arguments);
  };
}();

exports.estimateOsmoFee = estimateOsmoFee;

var noDecimals = function noDecimals(num) {
  return new _unit.IntPretty(new _unit.Dec(num)).maxDecimals(0).locale(false).toString();
};

exports.noDecimals = noDecimals;
