// osmojs/main/codegen/osmosis/concentrated-liquidity/position.js
"use strict";

var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PositionWithUnderlyingAssetBreakdown = exports.Position = void 0;
var _timestamp = require("../../google/protobuf/timestamp");
var _coin = require("../../cosmos/base/v1beta1/coin");
var _helpers = require("../../helpers");
var _m0 = _interopRequireWildcard(require("protobufjs/minimal"));
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function createBasePosition() {
  return {
    positionId: _helpers.Long.UZERO,
    address: "",
    poolId: _helpers.Long.UZERO,
    lowerTick: _helpers.Long.ZERO,
    upperTick: _helpers.Long.ZERO,
    joinTime: undefined,
    liquidity: ""
  };
}
var Position = {
  typeUrl: "/osmosis.concentratedliquidity.v1beta1.Position",
  encode: function encode(message) {
    var writer = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : _m0.Writer.create();
    if (!message.positionId.isZero()) {
      writer.uint32(8).uint64(message.positionId);
    }
    if (message.address !== "") {
      writer.uint32(18).string(message.address);
    }
    if (!message.poolId.isZero()) {
      writer.uint32(24).uint64(message.poolId);
    }
    if (!message.lowerTick.isZero()) {
      writer.uint32(32).int64(message.lowerTick);
    }
    if (!message.upperTick.isZero()) {
      writer.uint32(40).int64(message.upperTick);
    }
    if (message.joinTime !== undefined) {
      _timestamp.Timestamp.encode((0, _helpers.toTimestamp)(message.joinTime), writer.uint32(50).fork()).ldelim();
    }
    if (message.liquidity !== "") {
      writer.uint32(58).string(message.liquidity);
    }
    return writer;
  },
  decode: function decode(input, length) {
    var reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    var end = length === undefined ? reader.len : reader.pos + length;
    var message = createBasePosition();
    while (reader.pos < end) {
      var tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          message.positionId = reader.uint64();
          break;
        case 2:
          message.address = reader.string();
          break;
        case 3:
          message.poolId = reader.uint64();
          break;
        case 4:
          message.lowerTick = reader.int64();
          break;
        case 5:
          message.upperTick = reader.int64();
          break;
        case 6:
          message.joinTime = (0, _helpers.fromTimestamp)(_timestamp.Timestamp.decode(reader, reader.uint32()));
          break;
        case 7:
          message.liquidity = reader.string();
          break;
        default:
          reader.skipType(tag & 7);
          break;
      }
    }
    return message;
  },
  fromPartial: function fromPartial(object) {
    var _object$address, _object$joinTime, _object$liquidity;
    var message = createBasePosition();
    message.positionId = object.positionId !== undefined && object.positionId !== null ? _helpers.Long.fromValue(object.positionId) : _helpers.Long.UZERO;
    message.address = (_object$address = object.address) !== null && _object$address !== void 0 ? _object$address : "";
    message.poolId = object.poolId !== undefined && object.poolId !== null ? _helpers.Long.fromValue(object.poolId) : _helpers.Long.UZERO;
    message.lowerTick = object.lowerTick !== undefined && object.lowerTick !== null ? _helpers.Long.fromValue(object.lowerTick) : _helpers.Long.ZERO;
    message.upperTick = object.upperTick !== undefined && object.upperTick !== null ? _helpers.Long.fromValue(object.upperTick) : _helpers.Long.ZERO;
    message.joinTime = (_object$joinTime = object.joinTime) !== null && _object$joinTime !== void 0 ? _object$joinTime : undefined;
    message.liquidity = (_object$liquidity = object.liquidity) !== null && _object$liquidity !== void 0 ? _object$liquidity : "";
    return message;
  },
  fromAmino: function fromAmino(object) {
    return {
      positionId: _helpers.Long.fromString(object.position_id),
      address: object.address,
      poolId: _helpers.Long.fromString(object.pool_id),
      lowerTick: _helpers.Long.fromString(object.lower_tick),
      upperTick: _helpers.Long.fromString(object.upper_tick),
      joinTime: object !== null && object !== void 0 && object.join_time ? _timestamp.Timestamp.fromAmino(object.join_time) : undefined,
      liquidity: object.liquidity
    };
  },
  toAmino: function toAmino(message) {
    var obj = {};
    obj.position_id = message.positionId ? message.positionId.toString() : undefined;
    obj.address = message.address;
    obj.pool_id = message.poolId ? message.poolId.toString() : undefined;
    obj.lower_tick = message.lowerTick ? message.lowerTick.toString() : undefined;
    obj.upper_tick = message.upperTick ? message.upperTick.toString() : undefined;
    obj.join_time = message.joinTime ? _timestamp.Timestamp.toAmino(message.joinTime) : undefined;
    obj.liquidity = message.liquidity;
    return obj;
  },
  fromAminoMsg: function fromAminoMsg(object) {
    return Position.fromAmino(object.value);
  },
  toAminoMsg: function toAminoMsg(message) {
    return {
      type: "osmosis/concentratedliquidity/position",
      value: Position.toAmino(message)
    };
  },
  fromProtoMsg: function fromProtoMsg(message) {
    return Position.decode(message.value);
  },
  toProto: function toProto(message) {
    return Position.encode(message).finish();
  },
  toProtoMsg: function toProtoMsg(message) {
    return {
      typeUrl: "/osmosis.concentratedliquidity.v1beta1.Position",
      value: Position.encode(message).finish()
    };
  }
};
exports.Position = Position;
function createBasePositionWithUnderlyingAssetBreakdown() {
  return {
    position: undefined,
    asset0: undefined,
    asset1: undefined
  };
}
var PositionWithUnderlyingAssetBreakdown = {
  typeUrl: "/osmosis.concentratedliquidity.v1beta1.PositionWithUnderlyingAssetBreakdown",
  encode: function encode(message) {
    var writer = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : _m0.Writer.create();
    if (message.position !== undefined) {
      Position.encode(message.position, writer.uint32(10).fork()).ldelim();
    }
    if (message.asset0 !== undefined) {
      _coin.Coin.encode(message.asset0, writer.uint32(18).fork()).ldelim();
    }
    if (message.asset1 !== undefined) {
      _coin.Coin.encode(message.asset1, writer.uint32(26).fork()).ldelim();
    }
    return writer;
  },
  decode: function decode(input, length) {
    var reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    var end = length === undefined ? reader.len : reader.pos + length;
    var message = createBasePositionWithUnderlyingAssetBreakdown();
    while (reader.pos < end) {
      var tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          message.position = Position.decode(reader, reader.uint32());
          break;
        case 2:
          message.asset0 = _coin.Coin.decode(reader, reader.uint32());
          break;
        case 3:
          message.asset1 = _coin.Coin.decode(reader, reader.uint32());
          break;
        default:
          reader.skipType(tag & 7);
          break;
      }
    }
    return message;
  },
  fromPartial: function fromPartial(object) {
    var message = createBasePositionWithUnderlyingAssetBreakdown();
    message.position = object.position !== undefined && object.position !== null ? Position.fromPartial(object.position) : undefined;
    message.asset0 = object.asset0 !== undefined && object.asset0 !== null ? _coin.Coin.fromPartial(object.asset0) : undefined;
    message.asset1 = object.asset1 !== undefined && object.asset1 !== null ? _coin.Coin.fromPartial(object.asset1) : undefined;
    return message;
  },
  fromAmino: function fromAmino(object) {
    return {
      position: object !== null && object !== void 0 && object.position ? Position.fromAmino(object.position) : undefined,
      asset0: object !== null && object !== void 0 && object.asset0 ? _coin.Coin.fromAmino(object.asset0) : undefined,
      asset1: object !== null && object !== void 0 && object.asset1 ? _coin.Coin.fromAmino(object.asset1) : undefined
    };
  },
  toAmino: function toAmino(message) {
    var obj = {};
    obj.position = message.position ? Position.toAmino(message.position) : undefined;
    obj.asset0 = message.asset0 ? _coin.Coin.toAmino(message.asset0) : undefined;
    obj.asset1 = message.asset1 ? _coin.Coin.toAmino(message.asset1) : undefined;
    return obj;
  },
  fromAminoMsg: function fromAminoMsg(object) {
    return PositionWithUnderlyingAssetBreakdown.fromAmino(object.value);
  },
  toAminoMsg: function toAminoMsg(message) {
    return {
      type: "osmosis/concentratedliquidity/position-with-underlying-asset-breakdown",
      value: PositionWithUnderlyingAssetBreakdown.toAmino(message)
    };
  },
  fromProtoMsg: function fromProtoMsg(message) {
    return PositionWithUnderlyingAssetBreakdown.decode(message.value);
  },
  toProto: function toProto(message) {
    return PositionWithUnderlyingAssetBreakdown.encode(message).finish();
  },
  toProtoMsg: function toProtoMsg(message) {
    return {
      typeUrl: "/osmosis.concentratedliquidity.v1beta1.PositionWithUnderlyingAssetBreakdown",
      value: PositionWithUnderlyingAssetBreakdown.encode(message).finish()
    };
  }
};
exports.PositionWithUnderlyingAssetBreakdown = PositionWithUnderlyingAssetBreakdown;
