// osmojs/module/codegen/osmosis/concentrated-liquidity/tx.registry.js
//@ts-nocheck

import { MsgCreatePosition, MsgWithdrawPosition, MsgCollectFees, MsgCollectIncentives, MsgFungifyChargedPositions } from "./tx";
export const registry = [["/osmosis.concentratedliquidity.v1beta1.MsgCreatePosition", MsgCreatePosition], ["/osmosis.concentratedliquidity.v1beta1.MsgWithdrawPosition", MsgWithdrawPosition], ["/osmosis.concentratedliquidity.v1beta1.MsgCollectFees", MsgCollectFees], ["/osmosis.concentratedliquidity.v1beta1.MsgCollectIncentives", MsgCollectIncentives], ["/osmosis.concentratedliquidity.v1beta1.MsgFungifyChargedPositions", MsgFungifyChargedPositions]];
export const load = protoRegistry => {
  registry.forEach(([typeUrl, mod]) => {
    protoRegistry.register(typeUrl, mod);
  });
};
export const MessageComposer = {
  encoded: {
    createPosition(value) {
      return {
        typeUrl: "/osmosis.concentratedliquidity.v1beta1.MsgCreatePosition",
        value: MsgCreatePosition.encode(value).finish()
      };
    },
    withdrawPosition(value) {
      return {
        typeUrl: "/osmosis.concentratedliquidity.v1beta1.MsgWithdrawPosition",
        value: MsgWithdrawPosition.encode(value).finish()
      };
    },
    collectFees(value) {
      return {
        typeUrl: "/osmosis.concentratedliquidity.v1beta1.MsgCollectFees",
        value: MsgCollectFees.encode(value).finish()
      };
    },
    collectIncentives(value) {
      return {
        typeUrl: "/osmosis.concentratedliquidity.v1beta1.MsgCollectIncentives",
        value: MsgCollectIncentives.encode(value).finish()
      };
    },
    fungifyChargedPositions(value) {
      return {
        typeUrl: "/osmosis.concentratedliquidity.v1beta1.MsgFungifyChargedPositions",
        value: MsgFungifyChargedPositions.encode(value).finish()
      };
    }
  },
  withTypeUrl: {
    createPosition(value) {
      return {
        typeUrl: "/osmosis.concentratedliquidity.v1beta1.MsgCreatePosition",
        value
      };
    },
    withdrawPosition(value) {
      return {
        typeUrl: "/osmosis.concentratedliquidity.v1beta1.MsgWithdrawPosition",
        value
      };
    },
    collectFees(value) {
      return {
        typeUrl: "/osmosis.concentratedliquidity.v1beta1.MsgCollectFees",
        value
      };
    },
    collectIncentives(value) {
      return {
        typeUrl: "/osmosis.concentratedliquidity.v1beta1.MsgCollectIncentives",
        value
      };
    },
    fungifyChargedPositions(value) {
      return {
        typeUrl: "/osmosis.concentratedliquidity.v1beta1.MsgFungifyChargedPositions",
        value
      };
    }
  },
  fromPartial: {
    createPosition(value) {
      return {
        typeUrl: "/osmosis.concentratedliquidity.v1beta1.MsgCreatePosition",
        value: MsgCreatePosition.fromPartial(value)
      };
    },
    withdrawPosition(value) {
      return {
        typeUrl: "/osmosis.concentratedliquidity.v1beta1.MsgWithdrawPosition",
        value: MsgWithdrawPosition.fromPartial(value)
      };
    },
    collectFees(value) {
      return {
        typeUrl: "/osmosis.concentratedliquidity.v1beta1.MsgCollectFees",
        value: MsgCollectFees.fromPartial(value)
      };
    },
    collectIncentives(value) {
      return {
        typeUrl: "/osmosis.concentratedliquidity.v1beta1.MsgCollectIncentives",
        value: MsgCollectIncentives.fromPartial(value)
      };
    },
    fungifyChargedPositions(value) {
      return {
        typeUrl: "/osmosis.concentratedliquidity.v1beta1.MsgFungifyChargedPositions",
        value: MsgFungifyChargedPositions.fromPartial(value)
      };
    }
  }
};
