// osmojs/module/codegen/osmosis/incentives/params.js
import * as _m0 from "protobufjs/minimal";
/** Params holds parameters for the incentives module */

function createBaseParams() {
  return {
    distrEpochIdentifier: ""
  };
}
export const Params = {
  typeUrl: "/osmosis.incentives.Params",
  encode(message, writer = _m0.Writer.create()) {
    if (message.distrEpochIdentifier !== "") {
      writer.uint32(10).string(message.distrEpochIdentifier);
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseParams();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          message.distrEpochIdentifier = reader.string();
          break;
        default:
          reader.skipType(tag & 7);
          break;
      }
    }
    return message;
  },
  fromPartial(object) {
    var _object$distrEpochIde;
    const message = createBaseParams();
    message.distrEpochIdentifier = (_object$distrEpochIde = object.distrEpochIdentifier) !== null && _object$distrEpochIde !== void 0 ? _object$distrEpochIde : "";
    return message;
  },
  fromAmino(object) {
    return {
      distrEpochIdentifier: object.distr_epoch_identifier
    };
  },
  toAmino(message) {
    const obj = {};
    obj.distr_epoch_identifier = message.distrEpochIdentifier;
    return obj;
  },
  fromAminoMsg(object) {
    return Params.fromAmino(object.value);
  },
  toAminoMsg(message) {
    return {
      type: "osmosis/incentives/params",
      value: Params.toAmino(message)
    };
  },
  fromProtoMsg(message) {
    return Params.decode(message.value);
  },
  toProto(message) {
    return Params.encode(message).finish();
  },
  toProtoMsg(message) {
    return {
      typeUrl: "/osmosis.incentives.Params",
      value: Params.encode(message).finish()
    };
  }
};
