// osmojs/main/codegen/osmosis/rpc.query.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createRPCQueryClient = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _tendermintRpc = require("@cosmjs/tendermint-rpc");
var _stargate = require("@cosmjs/stargate");
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
var createRPCQueryClient = /*#__PURE__*/function () {
  var _ref2 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(_ref) {
    var rpcEndpoint, tmClient, client;
    return _regenerator["default"].wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            rpcEndpoint = _ref.rpcEndpoint;
            _context.next = 3;
            return _tendermintRpc.Tendermint34Client.connect(rpcEndpoint);
          case 3:
            tmClient = _context.sent;
            client = new _stargate.QueryClient(tmClient);
            _context.next = 7;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("../cosmos/auth/v1beta1/query.rpc.Query"));
            });
          case 7:
            _context.t0 = _context.sent.createRpcQueryExtension(client);
            _context.t1 = {
              v1beta1: _context.t0
            };
            _context.next = 11;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("../cosmos/authz/v1beta1/query.rpc.Query"));
            });
          case 11:
            _context.t2 = _context.sent.createRpcQueryExtension(client);
            _context.t3 = {
              v1beta1: _context.t2
            };
            _context.next = 15;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("../cosmos/bank/v1beta1/query.rpc.Query"));
            });
          case 15:
            _context.t4 = _context.sent.createRpcQueryExtension(client);
            _context.t5 = {
              v1beta1: _context.t4
            };
            _context.next = 19;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("../cosmos/base/node/v1beta1/query.rpc.Service"));
            });
          case 19:
            _context.t6 = _context.sent.createRpcQueryExtension(client);
            _context.t7 = {
              v1beta1: _context.t6
            };
            _context.t8 = {
              node: _context.t7
            };
            _context.next = 24;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("../cosmos/distribution/v1beta1/query.rpc.Query"));
            });
          case 24:
            _context.t9 = _context.sent.createRpcQueryExtension(client);
            _context.t10 = {
              v1beta1: _context.t9
            };
            _context.next = 28;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("../cosmos/gov/v1beta1/query.rpc.Query"));
            });
          case 28:
            _context.t11 = _context.sent.createRpcQueryExtension(client);
            _context.next = 31;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("../cosmos/gov/v1/query.rpc.Query"));
            });
          case 31:
            _context.t12 = _context.sent.createRpcQueryExtension(client);
            _context.t13 = {
              v1beta1: _context.t11,
              v1: _context.t12
            };
            _context.next = 35;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("../cosmos/staking/v1beta1/query.rpc.Query"));
            });
          case 35:
            _context.t14 = _context.sent.createRpcQueryExtension(client);
            _context.t15 = {
              v1beta1: _context.t14
            };
            _context.next = 39;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("../cosmos/tx/v1beta1/service.rpc.Service"));
            });
          case 39:
            _context.t16 = _context.sent.createRpcQueryExtension(client);
            _context.t17 = {
              v1beta1: _context.t16
            };
            _context.next = 43;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("../cosmos/upgrade/v1beta1/query.rpc.Query"));
            });
          case 43:
            _context.t18 = _context.sent.createRpcQueryExtension(client);
            _context.t19 = {
              v1beta1: _context.t18
            };
            _context.t20 = {
              auth: _context.t1,
              authz: _context.t3,
              bank: _context.t5,
              base: _context.t8,
              distribution: _context.t10,
              gov: _context.t13,
              staking: _context.t15,
              tx: _context.t17,
              upgrade: _context.t19
            };
            _context.next = 48;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./concentrated-liquidity/pool-model/query.rpc.Query"));
            });
          case 48:
            _context.t21 = _context.sent.createRpcQueryExtension(client);
            _context.t22 = {
              v1beta1: _context.t21
            };
            _context.next = 52;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./cosmwasmpool/v1beta1/query.rpc.Query"));
            });
          case 52:
            _context.t23 = _context.sent.createRpcQueryExtension(client);
            _context.t24 = {
              v1beta1: _context.t23
            };
            _context.next = 56;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./downtime-detector/v1beta1/query.rpc.Query"));
            });
          case 56:
            _context.t25 = _context.sent.createRpcQueryExtension(client);
            _context.t26 = {
              v1beta1: _context.t25
            };
            _context.next = 60;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./epochs/query.rpc.Query"));
            });
          case 60:
            _context.t27 = _context.sent.createRpcQueryExtension(client);
            _context.t28 = {
              v1beta1: _context.t27
            };
            _context.next = 64;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./gamm/v1beta1/query.rpc.Query"));
            });
          case 64:
            _context.t29 = _context.sent.createRpcQueryExtension(client);
            _context.next = 67;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./gamm/v2/query.rpc.Query"));
            });
          case 67:
            _context.t30 = _context.sent.createRpcQueryExtension(client);
            _context.t31 = {
              v1beta1: _context.t29,
              v2: _context.t30
            };
            _context.next = 71;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./ibc-rate-limit/v1beta1/query.rpc.Query"));
            });
          case 71:
            _context.t32 = _context.sent.createRpcQueryExtension(client);
            _context.t33 = {
              v1beta1: _context.t32
            };
            _context.next = 75;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./incentives/query.rpc.Query"));
            });
          case 75:
            _context.t34 = _context.sent.createRpcQueryExtension(client);
            _context.next = 78;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./lockup/query.rpc.Query"));
            });
          case 78:
            _context.t35 = _context.sent.createRpcQueryExtension(client);
            _context.next = 81;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./mint/v1beta1/query.rpc.Query"));
            });
          case 81:
            _context.t36 = _context.sent.createRpcQueryExtension(client);
            _context.t37 = {
              v1beta1: _context.t36
            };
            _context.next = 85;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./pool-incentives/v1beta1/query.rpc.Query"));
            });
          case 85:
            _context.t38 = _context.sent.createRpcQueryExtension(client);
            _context.t39 = {
              v1beta1: _context.t38
            };
            _context.next = 89;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./poolmanager/v1beta1/query.rpc.Query"));
            });
          case 89:
            _context.t40 = _context.sent.createRpcQueryExtension(client);
            _context.t41 = {
              v1beta1: _context.t40
            };
            _context.next = 93;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./protorev/v1beta1/query.rpc.Query"));
            });
          case 93:
            _context.t42 = _context.sent.createRpcQueryExtension(client);
            _context.t43 = {
              v1beta1: _context.t42
            };
            _context.next = 97;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./superfluid/query.rpc.Query"));
            });
          case 97:
            _context.t44 = _context.sent.createRpcQueryExtension(client);
            _context.next = 100;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./tokenfactory/v1beta1/query.rpc.Query"));
            });
          case 100:
            _context.t45 = _context.sent.createRpcQueryExtension(client);
            _context.t46 = {
              v1beta1: _context.t45
            };
            _context.next = 104;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./twap/v1beta1/query.rpc.Query"));
            });
          case 104:
            _context.t47 = _context.sent.createRpcQueryExtension(client);
            _context.t48 = {
              v1beta1: _context.t47
            };
            _context.next = 108;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./txfees/v1beta1/query.rpc.Query"));
            });
          case 108:
            _context.t49 = _context.sent.createRpcQueryExtension(client);
            _context.t50 = {
              v1beta1: _context.t49
            };
            _context.next = 112;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./valset-pref/v1beta1/query.rpc.Query"));
            });
          case 112:
            _context.t51 = _context.sent.createRpcQueryExtension(client);
            _context.t52 = {
              v1beta1: _context.t51
            };
            _context.t53 = {
              concentratedliquidity: _context.t22,
              cosmwasmpool: _context.t24,
              downtimedetector: _context.t26,
              epochs: _context.t28,
              gamm: _context.t31,
              ibcratelimit: _context.t33,
              incentives: _context.t34,
              lockup: _context.t35,
              mint: _context.t37,
              poolincentives: _context.t39,
              poolmanager: _context.t41,
              protorev: _context.t43,
              superfluid: _context.t44,
              tokenfactory: _context.t46,
              twap: _context.t48,
              txfees: _context.t50,
              valsetpref: _context.t52
            };
            return _context.abrupt("return", {
              cosmos: _context.t20,
              osmosis: _context.t53
            });
          case 116:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));
  return function createRPCQueryClient(_x) {
    return _ref2.apply(this, arguments);
  };
}();
exports.createRPCQueryClient = createRPCQueryClient;
