// osmojs/main/codegen/ibc/applications/fee/v1/tx.rpc.msg.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MsgClientImpl = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _m0 = _interopRequireWildcard(require("protobufjs/minimal"));
var _tx = require("./tx");
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
var MsgClientImpl = /*#__PURE__*/function () {
  function MsgClientImpl(rpc) {
    (0, _classCallCheck2["default"])(this, MsgClientImpl);
    (0, _defineProperty2["default"])(this, "rpc", void 0);
    this.rpc = rpc;
    this.registerPayee = this.registerPayee.bind(this);
    this.registerCounterpartyPayee = this.registerCounterpartyPayee.bind(this);
    this.payPacketFee = this.payPacketFee.bind(this);
    this.payPacketFeeAsync = this.payPacketFeeAsync.bind(this);
  }
  (0, _createClass2["default"])(MsgClientImpl, [{
    key: "registerPayee",
    value: function registerPayee(request) {
      var data = _tx.MsgRegisterPayee.encode(request).finish();
      var promise = this.rpc.request("ibc.applications.fee.v1.Msg", "RegisterPayee", data);
      return promise.then(function (data) {
        return _tx.MsgRegisterPayeeResponse.decode(new _m0.Reader(data));
      });
    }
  }, {
    key: "registerCounterpartyPayee",
    value: function registerCounterpartyPayee(request) {
      var data = _tx.MsgRegisterCounterpartyPayee.encode(request).finish();
      var promise = this.rpc.request("ibc.applications.fee.v1.Msg", "RegisterCounterpartyPayee", data);
      return promise.then(function (data) {
        return _tx.MsgRegisterCounterpartyPayeeResponse.decode(new _m0.Reader(data));
      });
    }
  }, {
    key: "payPacketFee",
    value: function payPacketFee(request) {
      var data = _tx.MsgPayPacketFee.encode(request).finish();
      var promise = this.rpc.request("ibc.applications.fee.v1.Msg", "PayPacketFee", data);
      return promise.then(function (data) {
        return _tx.MsgPayPacketFeeResponse.decode(new _m0.Reader(data));
      });
    }
  }, {
    key: "payPacketFeeAsync",
    value: function payPacketFeeAsync(request) {
      var data = _tx.MsgPayPacketFeeAsync.encode(request).finish();
      var promise = this.rpc.request("ibc.applications.fee.v1.Msg", "PayPacketFeeAsync", data);
      return promise.then(function (data) {
        return _tx.MsgPayPacketFeeAsyncResponse.decode(new _m0.Reader(data));
      });
    }
  }]);
  return MsgClientImpl;
}();
exports.MsgClientImpl = MsgClientImpl;
