// osmojs/main/codegen/ibc/applications/fee/v1/tx.registry.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.registry = exports.load = exports.MessageComposer = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _tx = require("./tx");
//@ts-nocheck

var registry = [["/ibc.applications.fee.v1.MsgRegisterPayee", _tx.MsgRegisterPayee], ["/ibc.applications.fee.v1.MsgRegisterCounterpartyPayee", _tx.MsgRegisterCounterpartyPayee], ["/ibc.applications.fee.v1.MsgPayPacketFee", _tx.MsgPayPacketFee], ["/ibc.applications.fee.v1.MsgPayPacketFeeAsync", _tx.MsgPayPacketFeeAsync]];
exports.registry = registry;
var load = function load(protoRegistry) {
  registry.forEach(function (_ref) {
    var _ref2 = (0, _slicedToArray2["default"])(_ref, 2),
      typeUrl = _ref2[0],
      mod = _ref2[1];
    protoRegistry.register(typeUrl, mod);
  });
};
exports.load = load;
var MessageComposer = {
  encoded: {
    registerPayee: function registerPayee(value) {
      return {
        typeUrl: "/ibc.applications.fee.v1.MsgRegisterPayee",
        value: _tx.MsgRegisterPayee.encode(value).finish()
      };
    },
    registerCounterpartyPayee: function registerCounterpartyPayee(value) {
      return {
        typeUrl: "/ibc.applications.fee.v1.MsgRegisterCounterpartyPayee",
        value: _tx.MsgRegisterCounterpartyPayee.encode(value).finish()
      };
    },
    payPacketFee: function payPacketFee(value) {
      return {
        typeUrl: "/ibc.applications.fee.v1.MsgPayPacketFee",
        value: _tx.MsgPayPacketFee.encode(value).finish()
      };
    },
    payPacketFeeAsync: function payPacketFeeAsync(value) {
      return {
        typeUrl: "/ibc.applications.fee.v1.MsgPayPacketFeeAsync",
        value: _tx.MsgPayPacketFeeAsync.encode(value).finish()
      };
    }
  },
  withTypeUrl: {
    registerPayee: function registerPayee(value) {
      return {
        typeUrl: "/ibc.applications.fee.v1.MsgRegisterPayee",
        value: value
      };
    },
    registerCounterpartyPayee: function registerCounterpartyPayee(value) {
      return {
        typeUrl: "/ibc.applications.fee.v1.MsgRegisterCounterpartyPayee",
        value: value
      };
    },
    payPacketFee: function payPacketFee(value) {
      return {
        typeUrl: "/ibc.applications.fee.v1.MsgPayPacketFee",
        value: value
      };
    },
    payPacketFeeAsync: function payPacketFeeAsync(value) {
      return {
        typeUrl: "/ibc.applications.fee.v1.MsgPayPacketFeeAsync",
        value: value
      };
    }
  },
  fromPartial: {
    registerPayee: function registerPayee(value) {
      return {
        typeUrl: "/ibc.applications.fee.v1.MsgRegisterPayee",
        value: _tx.MsgRegisterPayee.fromPartial(value)
      };
    },
    registerCounterpartyPayee: function registerCounterpartyPayee(value) {
      return {
        typeUrl: "/ibc.applications.fee.v1.MsgRegisterCounterpartyPayee",
        value: _tx.MsgRegisterCounterpartyPayee.fromPartial(value)
      };
    },
    payPacketFee: function payPacketFee(value) {
      return {
        typeUrl: "/ibc.applications.fee.v1.MsgPayPacketFee",
        value: _tx.MsgPayPacketFee.fromPartial(value)
      };
    },
    payPacketFeeAsync: function payPacketFeeAsync(value) {
      return {
        typeUrl: "/ibc.applications.fee.v1.MsgPayPacketFeeAsync",
        value: _tx.MsgPayPacketFeeAsync.fromPartial(value)
      };
    }
  }
};
exports.MessageComposer = MessageComposer;
